0.0.81.argocd/pkg/argocd/api.go

```diff
@@ -0,0 +1,271 @@
+package argocd
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/Masterminds/semver"
+	"github.com/miekg/dns"
+	"github.com/oslokommune/okctl-upgrade/0.0.81.argocd/pkg/logger"
+	"github.com/oslokommune/okctl/pkg/api"
+	"github.com/oslokommune/okctl/pkg/apis/okctl.io/v1alpha1"
+	"github.com/oslokommune/okctl/pkg/cfn"
+	"github.com/oslokommune/okctl/pkg/client"
+	"github.com/oslokommune/okctl/pkg/client/core"
+	"github.com/oslokommune/okctl/pkg/config/constant"
+	"github.com/oslokommune/okctl/pkg/controller/common/reconciliation"
+	"github.com/oslokommune/okctl/pkg/helm/charts/argocd"
+	"strings"
+)
+
+const (
+	argoCDNamespace      = "argocd"
+	argoCDDeploymentName = "argocd-application-controller"
+	argoCDContainerName  = "application-controller"
+
+	argoClientSecretName = "argocd/client_secret" //nolint:gosec
+	argoSecretKeyName    = "argocd/secret_key"    //nolint:gosec
+	argoSecretName       = "argocd-secret"
+	argoPrivateKeyName   = "argocd-privatekey"
+)
+
+var (
+	expectedContainerImagePreUpgradeVersion = semver.MustParse("1.7.2")
+	targetContainerImageVersion             = semver.MustParse("2.1.7")
+)
+
+// ArgoCD is a sample okctl component
+type ArgoCD struct {
+	log     logger.Logger
+	dryRun  bool
+	confirm bool
+	kubectl Kubectl
+}
+
+var errNothingToDo = errors.New("nothing to do")
+
+// Upgrade upgrades the component
+func (a ArgoCD) Upgrade() error {
+	a.log.Info("Upgrading ArgoCD")
+
+	err := a.preflight()
+	if err != nil {
+		if errors.Is(err, errNothingToDo) {
+			return nil
+		}
+
+		return fmt.Errorf("running preflight checks: %w", err)
+	}
+
+	o, err := initializeOkctl()
+	if err != nil {
+		return fmt.Errorf("initializing: %w", err)
+	}
+
+	clusterID := getClusterID(o)
+	state := o.StateHandlers(o.StateNodes())
+
+	services, err := o.ClientServices(state)
+	if err != nil {
+		return err
+	}
+
+	err = a.partiallyRemoveArgoCD(services, clusterID)
+	if err != nil {
+		return fmt.Errorf("removing ArgoCD: %w", err)
+	}
+
+	err = a.createArgoCD(services, clusterID, state, o.Declaration)
+	if err != nil {
+		return fmt.Errorf("creating ArgoCD: %w", err)
+	}
+
+	err = a.postflight()
+	if err != nil {
+		return fmt.Errorf("running postflight checks: %w", err)
+	}
+
+	a.log.Info("Upgrading ArgoCD done!")
+
+	return nil
+}
+
+func (a ArgoCD) preflight() error {
+	isInstalled, err := a.kubectl.hasDeployment(argoCDNamespace, argoCDDeploymentName)
+	if err != nil {
+		return fmt.Errorf("getting argocd: %w", err)
+	}
+
+	if !isInstalled {
+		a.log.Info("ArgoCD is not installed, not doing anything")
+		return nil
```

Comment:
We should probably abort the upgrade when ArgoCD is missing. User probably set `integrations.argoCD: false`

Reply:
Ops, that's a bug, good catch. Fixed.